File: backend/app/models/product.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        autoincrement=True,
    )

    slug: Mapped[str] = mapped_column(
        String(100),
        unique=True,
        nullable=False,
        index=True,
    )

    name: Mapped[str] = mapped_column(
        String(150),
        nullable=False,
    )

    category: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        index=True,
    )

    description: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )

    base_price: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    sale_price: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    badge: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    template_key: Mapped[str] = mapped_column(
        String(100),
        unique=True,
        nullable=False,
    )

    is_featured: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
        server_default="false",
    )

    is_active: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
        server_default="true",
    )

    display_order: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

File: frontend/types/product.ts
export type ProductCategory =
  | "romance"
  | "celebration"
  | "gratitude"
  | "friendship"
  | "family"
  | "memories";

export type Product = {
  id: number;
  slug: string;
  name: string;
  category: ProductCategory;
  description: string;
  basePrice: number;
  salePrice: number;
  badge: string | null;
  templateKey: string;
  featured: boolean;
  displayOrder: number;
};

export type ProductApiResponse = {
  id: number;
  slug: string;
  name: string;
  category: ProductCategory;
  description: string;
  base_price: number;
  sale_price: number;
  badge: string | null;
  template_key: string;
  is_featured: boolean;
  display_order: number;
  created_at: string;
  updated_at: string;
};